src/operator/tensor/elemwise_unary_op.h

```diff
@@ -31,173 +31,381 @@
+      const TBlob dest_blob = dest->aux_data(i);
+      CopyBlob<xpu>(s, &dest_blob, reqi, src_blob);
+    }
+  }
 
+  /*! \brief Generic copy NDArray */
+  template<typename xpu>
+  static inline void CopyNDArray(mshadow::Stream<xpu> *s,
+                                 const NDArray *dest,
+                                 const OpReqType reqi,
+                                 const NDArray& src) {
+    DCHECK_NE(dest->storage_type(), kDefaultStorage);
+    DCHECK_EQ(dest->storage_type(), src.storage_type());
+    AllocateGeometry(dest, &src);
+    CopyGeometryBlobs(s, dest, reqi, src);
+    CopyBlob(s, &dest->data(), reqi, src.data());
+  }
 
-template<typename xpu>
-void IdentityCompute(const nnvm::NodeAttrs& attrs,
+  /*! \brief Map NDArray vectors to TBlob vectors and pass to compute function */
+  template<typename xpu, typename FComputer>
+  static inline void MapToFCompute(const nnvm::NodeAttrs &attrs,
+                                   const OpContext &ctx,
+                                   const std::vector<NDArray> &inputs,
+                                   const std::vector<OpReqType> &req,
+                                   const std::vector<NDArray> &outputs,
+                                   FComputer computer) {
+    std::vector<TBlob> in_blobs, out_blobs;
+    in_blobs.reserve(inputs.size());
+    out_blobs.reserve(outputs.size());
+    for (size_t i = 0, n = inputs.size(); i < n; ++i) {
+      in_blobs.emplace_back(std::move(inputs[i].data()));
+    }
+    for (size_t i = 0, n = outputs.size(); i < n; ++i) {
+      out_blobs.emplace_back(std::move(outputs[i].data()));
+    }
+    computer(attrs, ctx, in_blobs, req, out_blobs);
+  }
+
+  /*! \brief Keep row shape[0] dimension and gather the remaining dimensions in location shape[1] */
+  template<typename DType, typename xpu>
+  static inline mshadow::Tensor<xpu, 2, DType> AsRowise2D(mshadow::Stream<xpu> *s,
+                                                          const TBlob& blob) {
+    const size_t dim = blob.shape_.ndim();
+    if (dim) {
+      TShape shape({blob.shape_[0], 1});
+      for (size_t i = 1; i < dim; ++i) {
+        shape[1] *= blob.shape_[i];
+      }
+      return mshadow::Tensor<xpu, 2, DType>(
+        blob.dptr<DType>(), mshadow::Shape2(shape[0], shape[1]), s);
+    }
+    return mshadow::Tensor<xpu, 2, DType>();
+  }
+};  // OpBase
+
+/*! \brief Unary operator class */
+class UnaryOp : public OpBase {
+  /*! \brief Infer the output storage geometry */
+  template<int n_in, int n_out>
+  static bool InitStorageGeometry(const nnvm::NodeAttrs& attrs,
+                                  const std::vector<NDArray>& inputs,
+                                  const std::vector<NDArray>& outputs) {
+    CHECK_EQ(inputs.size(), static_cast<size_t>(n_in))
+      << " in operator " << attrs.name;
+    CHECK_EQ(outputs.size(), static_cast<size_t>(n_out))
+      << " in operator " << attrs.name;
+    CHECK(n_in > 0 && n_out > 0);
```

Comment:
Should static_assert be used instead since n_in and n_out are known at compile time?